### PR TITLE
Copter: encapsulate auto takeoff into an ojbect

### PR DESCRIPTION
**Pull request by peterbarker**
similar to the encapsulation of "user takeoff" into an object

```
Board               AP_Periph  blimp  bootloader  copter  heli  iofirmware  plane  rover  sub
Durandal                       *      *           -88     -80               *      *      *
HerePro             *                                                                     
Hitec-Airspeed      *                                                                     
KakuteH7-bdshot                *      *           -72     -72               *      *      *
MatekF405                      *      *           -64     -64               *      *      *
Pixhawk1-1M-bdshot             *                  -104    -104              *      *      *
f103-QiotekPeriph   *                                                                     
f303-Universal      *                                                                     
iomcu                                                           *                         
revo-mini                      *      *           -88     -88               *      *      *
skyviper-v2450                                    -64                                     
```

### CODE REVIEW

**Comment by rmackay9** [comment]
Sorry, I haven't had time to review this but hopefully I will before I go off on holidays later this week.

File: ArduCopter/takeoff.cpp [hunk 1]
```diff
@@ -118,20 +112,27 @@ void Mode::_TakeOff::do_pilot_takeoff(float& pilot_climb_rate_cm)
...
+    const auto &wp_nav = copter.wp_nav;
+    auto *motors = copter.motors;
+    auto *pos_control = copter.pos_control;
+    auto *attitude_control = copter.attitude_control;
+
     // if not armed set throttle to zero and exit immediately
     if (!motors->armed() || !copter.ap.auto_armed) {
         // do not spool down tradheli when on the ground with motor interlock enabled
-        make_safe_ground_handling(copter.is_tradheli() && motors->get_interlock());
+        copter.flightmode->make_safe_ground_handling(copter.is_tradheli() && copter.motors->get_interlock());
```
Comment (IamPete1) [inline ArduCopter/takeoff.cpp:127]:
```suggestion
        copter.flightmode->make_safe_ground_handling(copter.is_tradheli() && motors->get_interlock());
```

**Review by IamPete1 — COMMENTED** [review]
LGTM, I can't see any change in behavior, just namespace stuff.

File: ArduCopter/mode.h [hunk 2]
```diff
@@ -8,6 +8,27 @@ class ParametersG2;
 
 class GCS_Copter;
 
+// object shared by both Guided and Auto for takeoff.
+// position controller controls vehicle but the user can control the yaw.
+class _AutoTakeoff {
```
Comment (IamPete1) [inline ArduCopter/mode.h:13]: Why not keep this class inside mode?

**Comment by IamPete1** [comment]
This is good to merge once that stray `copter.motors` is fixed.

**Comment by peterbarker** [comment]
I've had a look at why this saves bytes; here's the elf-diff output for get_position, a trivial function affected by this PR:
```
// return takeoff final position if takeoff has completed successfully
bool _AutoTakeoff::get_position(Vector3p& _complete_pos)
{
    // only provide location if takeoff has completed
    if (!complete) {
        return false;
    }

    complete_pos = _complete_pos;
    return true;
}
```

image

My reading is that:
 - the compiler throws in a useless instruction in `mov r3, r1`
 - the compiler's loading of the boolean into a register, and loading the address if the Vector3 into a register is rather less efficient when those addresses are in the bss; the two words at the end of the disappeared function will be those addresses.  Those addresses will only be available at the link stage, so not available to the compiler, so it must load those addresses into a register and then use them (i.e. they can't be directly incorporated into the ARM instructions).  Whereas in the new code the addresses offsets can be taken in relation to the "this" pointer in r0....
 -

**Pull request merged**